pkg/controller/controller.go

```diff
@@ -0,0 +1,285 @@
+	// startNewServiceWatcher acquires the lock before accessing the map.
+	r.startNewServiceWatcher(clusterID, kubeConfig)
+}
+
+func (r *LightHouseController) OnUpdate(clusterID string, kubeConfig *rest.Config) {
+	klog.Infof("Updating cluster: %s", clusterID)
+	r.remoteClustersMutex.Lock()
+	defer r.remoteClustersMutex.Unlock()
+	r.removeServiceWatcher(clusterID)
+	r.startNewServiceWatcher(clusterID, kubeConfig)
+}
+
+func (r *LightHouseController) OnRemove(clusterID string) {
+	r.remoteClustersMutex.Lock()
+	defer r.remoteClustersMutex.Unlock()
+	r.removeServiceWatcher(clusterID)
+}
+
+func (r *LightHouseController) startNewServiceWatcher(clusterID string, kubeConfig *rest.Config) {
+	clientSet, err := r.newClientset(kubeConfig)
+	if err != nil {
+		klog.Errorf("error creating client set for cluster %s: %s", clusterID, err.Error())
+		return
+	}
+	serviceInformer := cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(options meta_v1.ListOptions) (runtime.Object, error) {
+				return clientSet.CoreV1().Services(meta_v1.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options meta_v1.ListOptions) (watch.Interface, error) {
+				return clientSet.CoreV1().Services(meta_v1.NamespaceAll).Watch(options)
+			},
+		},
+		&v1.Service{},
+		0,
+		cache.Indexers{},
+	)
+	remoteCluster := &remoteCluster{
+		stopCh:              make(chan struct{}),
+		clusterID:           clusterID,
+		serviceInformer:     serviceInformer,
+		queue:               workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		remoteServices:      r.remoteServices,
+		federator:           r.federator,
+		remoteServicesMutex: &sync.Mutex{},
+	}
+	r.remoteClustersMutex.Lock()
+	defer r.remoteClustersMutex.Unlock()
+	r.remoteClusters[clusterID] = remoteCluster
+
+	serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				remoteCluster.queue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err == nil {
+				remoteCluster.queue.Add(key)
+			}
+		},
+	})
+
+	go remoteCluster.run()
+}
+
+func (r *LightHouseController) removeServiceWatcher(clusterID string) {
+	remoteCluster, ok := r.remoteClusters[clusterID]
+	if ok {
+		klog.V(2).Infof("Stopping watcher for %s", clusterID)
+		remoteCluster.close()
+		delete(r.remoteClusters, clusterID)
+	}
+}
+
+func (r *remoteCluster) close() {
+	close(r.stopCh)
+	r.queue.ShutDown()
+	klog.Infof("Watcher queue shutdown for cluster: %s", r.clusterID)
+}
+
+func (r *remoteCluster) run() {
+	defer utilruntime.HandleCrash()
+	go r.serviceInformer.Run(r.stopCh)
+	go r.runWorker()
+}
+
+func (r *remoteCluster) runWorker() {
+	for {
+		keyObj, shutdown := r.queue.Get()
+		if shutdown {
+			klog.Infof("Lighthouse watcher for cluster \"%q\" stopped", r.clusterID)
+			return
+		}
+		key := keyObj.(string)
+		func() {
```

Comment:
why do you create a lambda function, and just call it down? L208

This can be remove.

Reply:
This is done so `r.queue.Done(key)` can be deferred as it needs to be called on every path. Of course the code could be re-structured w/o the func as well. I can push another PR to change that - is that OK?

Reply:
I think it'd work without the func also, right?

I thought defer was handled at the end of a { .... }  block, Is it function level?

I'll merge.

Reply:
> I thought defer was handled at the end of a { .... } block, Is it function level?
> 
yes it is